src/Components/QuickGrid/src/Microsoft.AspNetCore.Components.QuickGrid/Infrastructure/AsyncQueryExecutorSupplier.cs

```diff
@@ -0,0 +1,56 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Components.QuickGrid.Infrastructure;
+
+internal static class AsyncQueryExecutorSupplier
+{
+    // The primary goal with this is to ensure that:
+    //  - If you're using EF Core, then we resolve queries efficiently using its ToXyzAsync async extensions and don't
+    //    just fall back on the synchronous IQueryable ToXyz calls
+    //  - ... but without QuickGrid referencing Microsoft.EntityFramework directly. That's because it would bring in
+    //    heavy dependencies you may not be using (and relying on trimming isn't enough, as it's still desirable to have
+    //    heavy unused dependencies for Blazor Server).
+    //
+    // As a side-effect, we have an abstraction IAsyncQueryExecutor that developers could use to plug in their own
+    // mechanism for resolving async queries from other data sources than EF. It's not really a major goal to make this
+    // adapter generally useful beyond EF, but fine if people do have their own uses for it.
+
+    private static readonly ConcurrentDictionary<Type, bool> IsEntityFrameworkProviderTypeCache = new();
+
+    public static IAsyncQueryExecutor? GetAsyncQueryExecutor<T>(IServiceProvider services, IQueryable<T>? queryable)
+    {
+        if (queryable is not null)
+        {
+            var executor = services.GetService<IAsyncQueryExecutor>();
+
+            if (executor is null)
+            {
+                // It's useful to detect if the developer is unaware that they should be using the EF adapter, otherwise
+                // they will likely never notice and simply deploy an inefficient app that blocks threads on each query.
+                var providerType = queryable.Provider?.GetType();
+                if (providerType is not null && IsEntityFrameworkProviderTypeCache.GetOrAdd(providerType, IsEntityFrameworkProviderType))
```

Comment:
The trimmer isn't a fan of this line because `IsEntityFrameworkProviderType` is accessed via reflection. @SteveSandersonMS what do you think about this?

Reply:
I know it's been a while since you raised this point so am not sure if any clarity has emerged since. Is this still an open question? TBH I don't know for sure what the right thing to do is, but would start by investigating if there are any linker annotations that match the scenario.

Reply:
I'll investigate. In the meantime, would you be willing to clarify the goals of this class? The comments are helpful but I admittedly don't see the full picture. If the type of your `IQueryable<T>` is supported by EF (or comes from EF?) and you're not using the adapter then an error is thrown?

Reply:
The goal is to be able to resolve queries asynchronously when possible. EF gives the developer an `IQueryable<T>` but they also give a set of async extension methods like `myQueryable.CountAsync()`. Those extension methods don't exist if you're not referencing EF and only work on EF-supplied IQueryables.

Since QuickGrid itself doesn't want5a direct dependency on EF (because in most cases it's not wanted in WebAssembly apps, and the linker wouldn't even be able to remove it if we're doing runtime checks to find out whether a given IQueryable comes from EF), the `Microsoft.AspNetCore.Components.QuickGrid.EntityFrameworkAdapter` package is an optional dependency that registers a DI service that does the runtime check ("is this IQueryable from EF?") and resolves QuickGrid's queries using the EF async extension methods instead of the synchronous IQueryable methods.

The particular line of code here (line 37) is where we've already checked if you have the optional adapter service and found that you do **not** (that's on lines 30-32). So the remaining purpose of this branch is to help the developer notice if they *should* have added the adapter service but forgot to do so. We are using reflection to check whether the current IQueryable is an EF-enabled one, because if it is, the developer has made a mistake in failing to register the adapter service.

As per the comment on lines 54-55, it doesn't matter if this check gets trimmed out in production or if it fails to notice EF-provided IQueryables, because it's only to help detect misconfiguration and is sufficient to do this in development only.

Reply:
Sounds good. In this case, it looks like we can suppress this warning. The runtime repo has tons of regression testing to ensure that what we'd want doesn't get trimmed out, but this doesn't seem like the time to look into setting it up in aspnetcore. Besides, the specific error (IL2111) is concerned with not guaranteeing meeting the `[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.Interfaces)]` annotation. We're okay with the prospect of an interface potentially being trimmed away because as you pointed out, this error is designed to be caught during development.